### PR TITLE
Refactor response classes

### LINKED ISSUE #2061: Refactor Kedro-Viz FAST API app creation
## Description

On Kedro-Viz we have a FAST API app in the backend and we use uvicorn web server to host the app. We register our routers and also install corresponding dependencies.

We include 2 routers before we start the app -
1. Register a REST API router and the required dependencies
2. Register a graphQL router and the required dependencies

We also do some expensive operations like SQLiteStore initialization, import modules like NetworkX, Azure related imports for DeployerFactory. 

These can slow down the app and increase the start time.

## Context

We need to refactor the app creation phase and also defer unwanted imports to improve the app start time.

## Possible Implementation

To improve `kedro viz run` time after `run` command is reached, we can explore and experiment with changes I started [here](https://github.com/kedro-org/kedro-viz/tree/chore/refactor-cli) based on the flame graph

Some of the observations made while working on improving performance for Kedro-Viz -

image

<hr>

- Module imports related to experiment tracking, strawberry graphQL (2-3sec) - We need to see if we can delay graphql call until someone clicks on ET, remove getVersions call (a middleware which ingests graphQL api app once a request for graphQL is made)
- Separate api app (main api + additional api) - This will help remove dependencies involved for other apis
- Module imports related to deployer factory (2-3sec) - Deferring deployer factory imports until a `/deploy` call is made
- Separate responses module and try to get the main route working quickly - This needs refactoring and breaking down our responses module.
- NetworkX takes time for import - Deferring the networkX import until used
- SQLiteStore initialization also takes time (reason for configure and bootstrap project taking time) - Need to find a way to defer store initialization (likely a North star)
- Apart from all the above improvements, `uvicorn.run` takes ~10sec to make the app available. Need to experiment by making a lighter app, reducing api routes and dependencies.

## Checklist

- [x] Include labels so that we can categorise your feature request

### PR DESCRIPTION
**Pull request by ravi-kumar-pilla**
## Description

Partially resolves #2061 

- [ ] Module imports related to experiment tracking, strawberry graphQL (2-3sec) - We need to see if we can delay graphql call until someone clicks on ET, remove getVersions call (a middleware which ingests graphQL api app once a request for graphQL is made)
- [ ] Separate api app (main api + additional api) - This will help remove dependencies involved for other apis
- [x] Module imports related to deployer factory (2-3sec) - Deferring deployer factory imports until a /deploy call is made
- [x] Separate responses module and try to get the main route working quickly - This needs refactoring and breaking down our responses module.
- [x] NetworkX takes time for import - Deferring the networkX import until used
- [ ] SQLiteStore initialization also takes time (reason for configure and bootstrap project taking time) - Need to find a way to defer store initialization (likely a North star)
- [x] Apart from all the above improvements, uvicorn.run takes ~10sec to make the app available. Need to experiment by making a lighter app, reducing api routes and dependencies.

## Development notes

* Split `responses.py` file and group the response classes under `responses` folder which helps in scaling and maintaining the responses code base
* Defer heavy imports like `NetworkX`, deployer factory imports until used
* By experimenting found the real reason for `uvicorn.run` taking time is app building which involves importing all kedro project modules (example - for demo project, sklearn, matplotlib etc take time as shown below).
* Updated tests and import references

image
 
**Next steps:**

* Moving away from multiprocessing to threading
* https://github.com/kedro-org/kedro-viz/issues/2079

## QA notes

* All tests should pass
* Since the responses module is refactored, there needs to be manual testing for confirming the CLI and UI responses. It would be nice if we test `kedro viz build`, `kedro viz deploy` are intact
* @jitu5 I think Kedro VSCode uses a function `get_kedro_project_json_data` in responses file which is moved to `pipelines.py` file. I would request you to have a look at this

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes

### CODE REVIEW

**Review by rashidakanchwala — COMMENTED** [review]
Hi, I had a think through this and came up with a few changes. Please let me know if it makes sense to you and happy to get on a call to discuss further. 

- save_responses.py file contains all the save_api and write_api functions. It's purpose it to handle saving API responses to FS. 

- utils. py file contains all encoding related functions classes + utility funcs

- node.py contains all response class for NodeAPI, and NodeMetadataAPI + get_node_metadata_response_function 

- pipeline.py contains all response classes for GraphAPI, NamedEntity, ModularPipelineResponses, and get_selected_pipeline_response function 

- metadata.py contains the same as you suggested. 

- base.py contains the BaseAPIResponse and APIErrorMesage response.

File: package/kedro_viz/api/rest/responses/save_responses.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
+"""`kedro_viz.api.rest.responses.save_responses` contains response classes
+and utility functions for the `/deploy` REST endpoint"""
```
Comment (rashidakanchwala) [inline package/kedro_viz/api/rest/responses/save_responses.py:2]: this is more than just `/deploy` right ? this is also used in kedro viz --save-file I think

File: package/tests/test_api/test_rest/test_responses/assert_helpers.py [hunk 2]
```diff
@@ -0,0 +1,544 @@
+import operator
```
Comment (rashidakanchwala) [inline package/tests/test_api/test_rest/test_responses/assert_helpers.py:1]: nice!!

File: package/kedro_viz/api/rest/responses/pipelines.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        layers=data_access_manager.get_sorted_layers_for_registered_pipeline(
+            pipeline_id
+        ),
+        pipelines=data_access_manager.registered_pipelines.as_list(),
+        modular_pipelines=modular_pipelines_tree,
+        selected_pipeline=pipeline_id,
+    )
+
+
+def get_kedro_project_json_data():
```
Comment (rashidakanchwala) [inline package/kedro_viz/api/rest/responses/pipelines.py:193]: was wondering if this can be moved to utils.py

**Review by rashidakanchwala — APPROVED** [review]
LGTM!! :) thank you this is a big refactor.  @ravi-kumar-pilla

File: package/kedro_viz/launchers/cli/utils.py [hunk 4]
```diff
@@ -104,16 +104,6 @@ def create_shareableviz_process(
         viz_deploy_process.terminate()
 
 
-def display_cli_message(msg, msg_color=None):
```
Comment (SajidAlamQB) [inline package/kedro_viz/launchers/cli/utils.py:107]: Why was this moved?

File: package/kedro_viz/api/rest/responses/nodes.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
+"""`kedro_viz.api.rest.responses.nodes` contains response classes
+and utility functions for the `/nodes/*` REST endpoints"""
+
+# pylint: disable=missing-class-docstring,invalid-name
```
Comment (SajidAlamQB) [inline package/kedro_viz/api/rest/responses/nodes.py:4]: We've switched to ruff these pylint commands will have to be updated.

**Review by SajidAlamQB — APPROVED** [review]
Hey @ravi-kumar-pilla , thanks for this PR the refactoring looks good to me, I like the way you've split the files! I've left some minor comments!

**Comment by jitu5** [comment]
@ravi-kumar-pilla 

> @jitu5 I think Kedro VSCode uses a function get_kedro_project_json_data in responses file which is moved to pipelines.py file. I would request you to have a look at this

Are we going with `pipelines.py` or `utils.py` for `get_kedro_project_json_data` ? 
To support for both Kdero-Viz version 10.0.0 and later in VSCode, I am thinking to add below

    try:
        from kedro_viz.api.rest.responses.utils import get_kedro_project_json_data
    except ImportError as e:
        from kedro_viz.api.rest.responses import get_kedro_project_json_data

File: package/kedro_viz/api/rest/responses/save_responses.py (hunk 1 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/api/rest/responses/save_responses.py:2]: I will update the text

File: package/kedro_viz/api/rest/responses/pipelines.py (hunk 3 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/api/rest/responses/pipelines.py:193]: Hi @rashidakanchwala, from the function name it seems to be a utility, but it is used to provide a pipeline response. Also we use `get_encoded_response` from utils. Placing this in utils will create a circular dependency.

File: package/kedro_viz/launchers/cli/utils.py (hunk 4 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/launchers/cli/utils.py:107]: Earlier this was used only in cli. In this PR I tried using it in dev mode in [`server.py`](https://github.com/kedro-org/kedro-viz/pull/2113/files#diff-096675efe9d411d7fb88ddc3a449afef347ba47fb5a6aa5812f8602738e240efR179) 
To avoid circular dependency, this needs to be moved

**Comment by ravi-kumar-pilla** [comment]
> Are we going with `pipelines.py` or `utils.py` for `get_kedro_project_json_data` ? To support for both Kdero-Viz version 10.0.0 and later in VSCode, I am thinking to add below
> 
> ```
> try:
>     from kedro_viz.api.rest.responses.utils import get_kedro_project_json_data
> except ImportError as e:
>     from kedro_viz.api.rest.responses import get_kedro_project_json_data    
> ```

Hi @jitu5 , yes it will be in pipelines.py as it is now
